### PR TITLE
incremental ededup

### PR DESCRIPTION
**Pull request by blublinsky**
## Why are these changes needed?
Currently ededup does not support incremental execution. 
This PR:
  1. Adds hash snapshotting to ededup execution (currently at the end)
  2. Supports loading known hashes, thus filtering new files based on the result

### CODE REVIEW

**Comment by blublinsky** [comment]
Only Python support is implemented. Ray to come

**Comment by blublinsky** [comment]
please do not merge yet

**Comment by blublinsky** [comment]
And now all set

File: transforms/universal/ededup/ray/README.md
Comment (daw3rd) [inline transforms/universal/ededup/ray/README.md:13]: Instead of duplicating what is in the python README, just reference it.  Of course, if there are ray-specific concifgurations, they should be listed here.  That way you won't have to make my requested changes in two places :).

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [inline transforms/universal/ededup/python/README.md]: The ededup_ prefix should only be needed/used for the CLI parameters.

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [inline transforms/universal/ededup/python/README.md]: I think you mean `snapshots` instead of `transforms` here.

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [inline transforms/universal/ededup/python/README.md:1]: I think. you need a discussion of snapshotting and why/when it should be used.

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: transforms/universal/ededup/python/README.md
Comment (blublinsky) [reply transforms/universal/ededup/python/README.md]: It is a cli parameter

Comment (blublinsky) [reply transforms/universal/ededup/python/README.md]: fixed

File: transforms/universal/ededup/python/README.md
Comment (blublinsky) [reply transforms/universal/ededup/python/README.md]: fixed

File: transforms/universal/ededup/python/README.md
Comment (blublinsky) [reply transforms/universal/ededup/python/README.md:1]: Added to documentation

File: data-processing-lib/python/src/data_processing/data_access/snapshot_utils.py
Comment (daw3rd) [inline data-processing-lib/python/src/data_processing/data_access/snapshot_utils.py:1]: who else is using this?  if no one. i don't think it belongs in the core library

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [inline transforms/universal/ededup/python/README.md]: defaults should be specified as well.  perhaps follow the table idea from here https://github.com/IBM/data-prep-kit/tree/dev/transforms/language/pdf2parquet/python#parameters

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [reply transforms/universal/ededup/python/README.md:1]: I don't see it in the python/README.md

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [inline transforms/universal/ededup/python/README.md]: can we not have multiple terms for the same thing.  How about incremental=True instead of use_snapshot=True.

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (daw3rd) [inline transforms/universal/ededup/python/src/ededup_transform_base.py]: again, please use a constant for these key values both here and in the parser setup.

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (daw3rd) [inline transforms/universal/ededup/python/src/ededup_transform_python.py]: I think we need a default here so the transform can more easily be run outside of the python runtime/launcher.

File: transforms/universal/ededup/python/test-data/incremental/snapshot/hash_collector_1
Comment (daw3rd) [inline transforms/universal/ededup/python/test-data/incremental/snapshot/hash_collector_1:1]: is this file really being validated? or should it be in .gitignore?

File: transforms/universal/ededup/python/test-data/input/snapshot/hash_collector_1
Comment (daw3rd) [inline transforms/universal/ededup/python/test-data/input/snapshot/hash_collector_1:1]: why was this created in the input dir?  i thought the snapshot dir was in the output.  Same comment about .gitignore.

File: transforms/universal/ededup/python/test/test_ededup_python_incremental.py
Comment (daw3rd) [inline transforms/universal/ededup/python/test/test_ededup_python_incremental.py]: Resize -> Ededup

Comment (daw3rd) [reply transforms/universal/ededup/python/test/test_ededup_python_incremental.py]: Instead of a whole new class, you should just be able to append the fixture tuple to the list of fixtures in the original test.

File: transforms/universal/ededup/ray/README.md
Comment (daw3rd) [inline transforms/universal/ededup/ray/README.md]: Ok, this sounds like a nice feature.  So I am guessing, but maybe you should clarify, that the 2nd set of input must use the same output directory so that it can pick up the snapshots from that 1st/shared output directory.  If this is the case, why not always enable this feature?

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: data-processing-lib/python/src/data_processing/data_access/snapshot_utils.py
Comment (blublinsky) [reply data-processing-lib/python/src/data_processing/data_access/snapshot_utils.py:1]: fuzzy dedup

File: transforms/universal/ededup/python/README.md
Comment (blublinsky) [reply transforms/universal/ededup/python/README.md]: Incremental is one case of using transforms

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_python.py]: there is no default. we can run it outside python. see example

File: transforms/universal/ededup/python/test-data/incremental/snapshot/hash_collector_1
Comment (blublinsky) [reply transforms/universal/ededup/python/test-data/incremental/snapshot/hash_collector_1:1]: It is validated. We make sure that output structure as expected

File: transforms/universal/ededup/python/test-data/input/snapshot/hash_collector_1
Comment (blublinsky) [reply transforms/universal/ededup/python/test-data/input/snapshot/hash_collector_1:1]: because its input for incremental

File: transforms/universal/ededup/python/test/test_ededup_python_incremental.py
Comment (blublinsky) [reply transforms/universal/ededup/python/test/test_ededup_python_incremental.py]: fixed

File: transforms/universal/ededup/ray/README.md
Comment (blublinsky) [reply transforms/universal/ededup/ray/README.md]: Its not. By default it picks it up from output directory, but you van overwrite it

File: transforms/universal/ededup/ray/README.md
Comment (blublinsky) [reply transforms/universal/ededup/ray/README.md:13]: fixed in both places

File: data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py
Comment (roytman) [inline data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py:138]: we use the string "processing_time" in 5 different places, can we replace it with a constant? 
P.S. it is relevant to outer strings.

File: transforms/universal/ededup/README.md
Comment (roytman) [inline transforms/universal/ededup/README.md]: enables running **a** ray docker image

File: transforms/universal/ededup/python/README.md
Comment (roytman) [inline transforms/universal/ededup/python/README.md]: I think the previous definition was clearer.

File: transforms/universal/ededup/python/src/ededup_local.py
Comment (roytman) [inline transforms/universal/ededup/python/src/ededup_local.py:15]: pre-commit complains about it. (according to ABC "d" is before "c")

File: data-processing-lib/python/src/data_processing/test_support/abstract_test.py
Comment (roytman) [inline data-processing-lib/python/src/data_processing/test_support/abstract_test.py]: please run pre-commit

File: transforms/universal/ededup/python/src/ededup_local_python.py
Comment (roytman) [inline transforms/universal/ededup/python/src/ededup_local_python.py]: string constants

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (roytman) [inline transforms/universal/ededup/python/src/ededup_transform_base.py:33]: Should the constants `short_name`  and `cli_prefix` be uppercase in the next lines?

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (roytman) [inline transforms/universal/ededup/python/src/ededup_transform_python.py:15]: please run `pre-commit`

**Review by roytman — COMMENTED** [review]
several small comments

File: data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py
Comment (blublinsky) [reply data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py:138]: Its in so many different files

File: transforms/universal/ededup/README.md
Comment (blublinsky) [reply transforms/universal/ededup/README.md]: done

File: transforms/universal/ededup/python/README.md
Comment (blublinsky) [reply transforms/universal/ededup/python/README.md]: fixed

File: transforms/universal/ededup/python/src/ededup_local.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_local.py:15]: run pre-commit

File: data-processing-lib/python/src/data_processing/test_support/abstract_test.py
Comment (blublinsky) [reply data-processing-lib/python/src/data_processing/test_support/abstract_test.py]: done

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_base.py:33]: no rule for that

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_python.py:15]: done

File: data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py
Comment (daw3rd) [reply data-processing-lib/ray/src/data_processing_ray/runtime/ray/transform_orchestrator.py:138]: which is why in general, we like constants :).  Let's please try and use them in the future.

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [reply transforms/universal/ededup/python/README.md:1]: To be consistent, the python implementation README holds the details, not ray.  Can you please move the details to python readme.

File: transforms/universal/ededup/ray/README.md
Comment (daw3rd) [inline transforms/universal/ededup/ray/README.md:32]: These are all duplicated from python.  Lets just link to the python CLI parameters.

File: transforms/universal/ededup/ray/README.md
Comment (daw3rd) [reply transforms/universal/ededup/ray/README.md]: This detail section is nice.  but maybe you can put it in its own file (in the python directory) and link to it from python README. Ray readme should mostly just defer/link to the python readme when details are involved.

File: transforms/universal/ededup/python/README.md
Comment (daw3rd) [reply transforms/universal/ededup/python/README.md]: i don't believe that speaks to my point.  having multiple terms is just plain obfuscation.  can we pick one please?

File: transforms/universal/ededup/python/src/ededup_local.py
Comment (daw3rd) [inline transforms/universal/ededup/python/src/ededup_local.py]: I think I've asked elsewhere, but can we have a default for "filter" at least in the python case?

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (daw3rd) [reply transforms/universal/ededup/python/src/ededup_transform_python.py]: But this requires the call to provide "filter". why should we required this when it seems we have an implementation that is expected to be used?

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (daw3rd) [inline transforms/universal/ededup/python/src/ededup_transform_python.py:89]: Is it ok to hardcode "hash_collector_1" and why the "_1"?  Perhaps some documentation on the structure of the snapshot directory would be useful - reading code to figure out is not nice.

File: transforms/universal/ededup/python/test/test_ededup_python_incremental.py
Comment (daw3rd) [reply transforms/universal/ededup/python/test/test_ededup_python_incremental.py]: I don't see either of these comments being addressed.

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: transforms/universal/ededup/python/README.md
Comment (blublinsky) [reply transforms/universal/ededup/python/README.md]: We are picking one - snapshotting. It is used for incremental, among other things

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_base.py]: Its going to be an import nightmare

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_python.py]: I still thing its wrong, but added this

File: transforms/universal/ededup/python/test/test_ededup_python_incremental.py
Comment (blublinsky) [reply transforms/universal/ededup/python/test/test_ededup_python_incremental.py]: Now the name is fixed. I wanted this to be a separate test

File: transforms/universal/ededup/ray/README.md
Comment (blublinsky) [reply transforms/universal/ededup/ray/README.md]: Restractured Readme's moving details into common and leaving only relevant for runtime info in runtimes

File: transforms/universal/ededup/ray/README.md
Comment (blublinsky) [reply transforms/universal/ededup/ray/README.md:32]: Not quite, they are different, there are additional parameters in Ray

File: transforms/universal/ededup/python/src/ededup_local.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_local.py]: Its done

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_python.py:89]: Because in the case of Python, there is only 1

File: transforms/universal/ededup/README.md
Comment (daw3rd) [inline transforms/universal/ededup/README.md:1]: by convention this file just points to the runtimes.  The detail is all good, but for consistency, most of this content should be in the python/README.md

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (daw3rd) [reply transforms/universal/ededup/python/src/ededup_transform_base.py]: no one is required to use the constants, but it would be highly recommended.  Mostly I suggest it for your own implementation internally.  Makes for better error checking and less delicate code.

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (daw3rd) [inline transforms/universal/ededup/python/src/ededup_transform_python.py]: This check is no longer needed.

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: transforms/universal/ededup/README.md
Comment (blublinsky) [reply transforms/universal/ededup/README.md:1]: Python references it. It really belongs outside of the runtime, as it belongs to both

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_python.py]: fixed

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_base.py]: done

File: transforms/universal/ededup/README.md
Comment (daw3rd) [reply transforms/universal/ededup/README.md:1]: i see your point, although when the python ray and spark implementations are different, it is best to defer to the runtime readme. just trying for consistency when possible.

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (daw3rd) [inline transforms/universal/ededup/python/src/ededup_transform_python.py]: should the logger be a self.logger like we now do for other transforms?

File: transforms/universal/ededup/ray/src/ededup_transform_ray.py
Comment (daw3rd) [inline transforms/universal/ededup/ray/src/ededup_transform_ray.py]: maybe this long bit of code under this if could go in a _load_snapshots() method

**Review by daw3rd — CHANGES_REQUESTED** [review]

File: transforms/universal/ededup/python/src/ededup_transform_base.py
Comment (roytman) [reply transforms/universal/ededup/python/src/ededup_transform_base.py:33]: there is a PEP8 recommendation to name constants with uppercase letters. 
https://peps.python.org/pep-0008/#constants

> Constants are usually defined on a module level and written in all capital letters with underscores separating words. Examples include MAX_OVERFLOW and TOTAL.

File: transforms/universal/ededup/python/src/ededup_transform_python.py
Comment (blublinsky) [reply transforms/universal/ededup/python/src/ededup_transform_python.py]: done

File: transforms/universal/ededup/ray/src/ededup_transform_ray.py
Comment (blublinsky) [reply transforms/universal/ededup/ray/src/ededup_transform_ray.py]: done

**Review by daw3rd — APPROVED** [review]

**Pull request merged**
